src/uopool/services/simulation.rs

```diff
@@ -403,6 +404,126 @@ impl<M: Middleware + 'static> UoPoolService<M> {
         Ok(())
     }
 
+    async fn get_code_hashes(
+        &self,
+        contract_addresses: Vec<Address>,
+        code_hashes: &mut Vec<CodeHash>,
+    ) -> Result<(), SimulateValidationError> {
+        let mut tasks: JoinSet<Option<(Address, H256)>> = JoinSet::new();
+
+        for contract_address in contract_addresses {
+            let eth_provider = self.eth_provider.clone();
+
+            tasks.spawn(async move {
+                match eth_provider.get_code(contract_address, None).await {
+                    Ok(code) => Some((contract_address, keccak256(&code).into())),
+                    Err(_) => None,
+                }
+            });
+        }
+
+        while let Some(result) = tasks.join_next().await {
+            match result {
+                Ok(Some(code_hash)) => code_hashes.push(CodeHash {
+                    address: code_hash.0,
+                    hash: code_hash.1,
+                }),
+                Ok(None) | Err(_) => {
+                    return Err(SimulateValidationError::UnknownError {
+                        error: "failed to get code hashes".to_string(),
+                    });
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn compare_code_hashes(
+        &self,
```

Comment:
`self` is acutally not needed here. We could make the compare into a more general method.